Tools/paletteable sprite editor/js/lib/marvinj-0.9.js

```diff
@@ -0,0 +1,2665 @@
+				parseFloat(attr[3]),
+				parseFloat(attr[4]),
+				parseFloat(attr[5]),
+				parseFloat(attr[6])
+			);
+			*/
+			
+			this.rules.push(r);
+		}
+	}
+	
+	IteratedFunctionSystem.prototype.getRule = function(){
+		var random = Math.random();
+		var sum=0;
+		var i;
+		for(i=0; i<this.rules.length; i++){
+			sum += this.rules[i].probability;
+			if(random < sum){
+				return this.rules[i];
+			}
+		}
+		
+		if(i != 0){
+			return this.rules[i-1];
+		}
+		return this.rules[0];
+	};
+	
+	IteratedFunctionSystem.prototype.applyRule = function(point, rule){
+		var nx = rule.a*point[0] + rule.b*point[1]+rule.e;
+		var ny = rule.c*point[0] + rule.d*point[1]+rule.f;
+		point[0] = nx;
+		point[1] = ny;
+	};
+
+	function Crop(){
+		MarvinAbstractImagePlugin.super(this);
+		this.load();
+	}
+
+	Crop.prototype.load = function(){
+		this.setAttribute("x", 0);
+		this.setAttribute("y", 0);
+		this.setAttribute("width", 0);
+		this.setAttribute("height", 0);
+	};
+	
+	Crop.prototype.process = function
+	(
+		imageIn, 
+		imageOut,
+		attributesOut,
+		mask, 
+		previewMode
+	)
+    {
+		var x = this.getAttribute("x");
+		var y = this.getAttribute("y");
+		var width = this.getAttribute("width");
+		var height = this.getAttribute("height");
+		
+		imageOut.setDimension(width, height);
+		
+		for(var i=x; i<x+width; i++){
+			for(var j=y; j<y+height; j++){
+				imageOut.setIntColor(i-x, j-y, imageIn.getIntColor(i, j));
+			}
+		}
+    };
+
+
+	function FloodfillSegmentation(){
+		MarvinAbstractImagePlugin.super(this);
+		this.load();
+	}
+
+	FloodfillSegmentation.prototype.load = function(){
+		this.setAttribute("returnType", "MarvinSegment");
+	};
+	
+	FloodfillSegmentation.prototype.process = function
+	(
+		imageIn, 
+		imageOut,
+		attributesOut,
+		mask, 
+		previewMode
+	)
+	{
+		if(attributesOut != null){
+			var returnType = this.getAttribute("returnType");
+			var fillBuffer = imageIn.clone();
+			var segments = this.floodfillSegmentation(imageIn, fillBuffer);
+			
+			switch(returnType){
+				case "MarvinSegment":
+					attributesOut.set("segments", segments);
+					break;
+				case "MarvinBlobSegment":
+					attributesOut.set("blobSegments", blobSegments(fillBuffer, segments));
+					break;
+			}
+		}
+	};
+	
+	FloodfillSegmentation.prototype.floodfillSegmentation = function(image, fillBuffer){
+		fillBuffer.clear(0xFF000000);
+		
+		var currentColor=1;
+		for(var y=0; y<image.getHeight(); y++){
+			for(var x=0; x<image.getWidth(); x++){
+				
+				var color = fillBuffer.getIntColor(x, y);
+				
+				if((color & 0x00FFFFFF) == 0 && image.getAlphaComponent(x, y) > 0){
+					var c = 0xFF000000 | (currentColor++);
+					Marvin.boundaryFill(image, fillBuffer, x, y, c);
+				}
+			}
+		}
+		
+		var segments = new Array(currentColor-1);
+		var seg;
+		for(var y=0; y<fillBuffer.getHeight(); y++){
+			for(var x=0; x<fillBuffer.getWidth(); x++){
+				var color = (fillBuffer.getIntColor(x, y) & 0x00FFFFFF);
+				
+				if(color != 0x00FFFFFF && color > 0){
```

Comment:
Codacy found an issue: [Use ===/!== to compare with true/false or Numbers](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6395467)